# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/vm/BytecodeUtil.cpp
# Commit: eeaeb40f8b86
# Full Hash: eeaeb40f8b869121eff65b1c2d11af423c9a9ecc
# Author: Iain Ireland <iireland@mozilla.com>
# Date: 2022-03-09 09:44:44
# Regressor Bug: 885514
# File Overlap Count: 1
# Description:
#   Bug 885514: Simplify JSOp::Retsub r=jandem
#   
#   Currently, JSOp::Retsub takes two arguments, and uses one to decide whether to throw the second or use it as a resume index. This patch splits out the exception-handling logic into explicit bytecode, leaving JSOp::Retsub as a simple indirect jump.
#   
#   To make the generated bytecode work out nicely, I've swapped the order of the arguments: now `throwing` is on top, and the exception/resumeindex argument is underneath it.
# ==============================================================================

diff -r 1cb0b352b024 -r eeaeb40f8b86 js/src/vm/BytecodeUtil.cpp
--- a/js/src/vm/BytecodeUtil.cpp	Tue Mar 08 21:10:22 2022 +0000
+++ b/js/src/vm/BytecodeUtil.cpp	Tue Mar 08 21:10:22 2022 +0000
@@ -945,10 +945,9 @@
       case JSOp::ResumeIndex: {
         // ResumeIndex is used to push a return address for a finally block. If
         // this op is reachable, then so is that return address (with a smaller
-        // stack depth because the resume index and `throwing` will have been
-        // popped.
+        // stack depth because the resume index will have been popped.
         uint32_t resumeOffset = script_->resumeOffsets()[GET_UINT24(pc)];
-        if (!recordBytecode(resumeOffset, offsetStack, stackDepth - 2)) {
+        if (!recordBytecode(resumeOffset, offsetStack, stackDepth - 1)) {
           return false;
         }
         break;
@@ -2116,10 +2115,10 @@
         // Used for the values live on entry to the finally block.
         // See TryNoteKind::Finally above.
         if (defIndex == 0) {
-          return write("THROWING");
+          return write("PC");
         }
         MOZ_ASSERT(defIndex == 1);
-        return write("PC");
+        return write("THROWING");
 
       case JSOp::FunctionThis:
       case JSOp::ImplicitThis: